add generalized dot product

dkarrasch opened the pull request:
I went ahead and added a generalized dot product `dot(x, A, y)`, as discussed in JuliaLang/LinearAlgebra.jl#334. I shamelessly copied (and slightly adjusted) @simonbyrne's code for the completely sparse case https://github.com/JuliaLang/LinearAlgebra.jl/issues/334 and the dense case https://github.com/JuliaLang/LinearAlgebra.jl/issues/334, and added methods for structured matrices and sparse matrices/generic vectors. There are presumably a couple of things to do:

- [x] add tests

- [x] NEWS entry (I'm not sure, but I guess we would want to announce this feature)

- [x] further optimizations (edit: I tried to reduce memory access similar to the corresponding multiplication functions, such that I hope that avoiding allocating the intermediate multiplication result will pay off)

I'm not an expert in `@simd` and these kind of enhancements, so I'm happy to learn whether one would expect improvements by using such features.

This is written with the option of block matrices in mind, and that should be tested obviously.~~Since we need `zero` of eltypes for type promotion, how would one do that? We don't have StaticArrays to our disposal.~~

If I'm not mistaken, then this would close JuliaLang/LinearAlgebra.jl#334. ~~I don't see how one would leverage symmetry of `A`, or the fact that `x === y`.~~

dkarrasch committed:
add generalized dot product

chethega commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -793,6 +793,30 @@ function dot(x::AbstractArray, y::AbstractArray)
+For complex vectors, the first vector is conjugated.
+"""
+function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+    m = checksquare(A)
```
Is there a reason to restrict to square matrices? We only need `x'*A*y` to make sense.

dkarrasch replied:
No reason, I was a bit mislead by the structured matrix cases, and obviously unable to count dimensions correctly. 😄

chethega commented:
I think we also want specializations for `Transpose` and `Adjoint` for better memory access patterns. 

For example `gendot(x, A::Adjoint, y) = gendot(y, A', x)'` and something for `gendot(x', a::Transpose, y)`. I think we should write out the code instead of trying to be clever like above, especially since we probably need to write something for the transpose case anyway. I have not thought about possible ambiguity errors yet, so don't trust the above signature.

chethega commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -793,6 +793,57 @@ function dot(x::AbstractArray, y::AbstractArray)
+        return dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
+    end
+
+    if m == n
```
These test the numeric values in the matrix A at runtime for various properties, if these are not encoded in the type. If returning true, then we have already iterated over the entire matrix, and hence already paid large parts of the cost of this function (i.e. memory transfers).

Such runtime tests for useful special properties make sense for expensive operations (e.g. matrix factorizations) where the number of floating point ops is much larger than the number of memory transfers. I don't think they make sense here.

dkarrasch replied:
I added them to trigger comments like these, either in favor or against. I knew there is some effort related to these checks, but wasn't sure how they would weigh against the benefits from using the specialized methods.

stevengj replied:
I agree that the checks for specialized matrix types should be omitted — in this function they are as expensive as the dot product itself.

dkarrasch replied:
This is removed now and forever.

stevengj commented:
> I think we also want specializations for Transpose and Adjoint for better memory access patterns.

This wouldn't hurt, but I'm not clear on how useful it would be since for any `dot(x, A, y)` that defines a true inner product you should have `A == A'`, so there's not much point in doing `dot(x, A', y)`.

dkarrasch commented:
@stevengj I had true inner products in mind when I had the square assumption earlier, but I guess we are implementing here the machinery for ternary multiplication in general, in which there is no square and no symmetry assumption?

stevengj commented:
Sure, the `dot(x, A, y)` function should *work* for general `A`.  I'm just not sure how much trouble we should take to *optimize* nonsymmetric cases.

stevengj commented:
Probably should also have a specialized implementation for `UniformScaling` so that `dot(x, I, y)` calls `dot(x,y)`

stevengj commented:
Probably should also have a specialized implementation for `UniformScaling` so that `dot(x, I, y)` calls `dot(x,y)`.

dkarrasch committed:
add generalized dot for Adjoint and Transpose

dkarrasch committed:
add "generalized" dot for UniformScalings

dkarrasch committed:
fix adjoint/transpose in tridiags

stevengj commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -808,38 +808,60 @@ function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+    end
+    return s
+end
+function dot(x::AbstractVector, adjA::Adjoint, y::AbstractVector)
```
Can't this just be
```jl
dot(x::AbstractVector, adjA::Adjoint, y::AbstractVector) = adjoint(dot(y, A', x))
```
?

dkarrasch replied:
Sure. So you want to not treat the non-real transpose case explicitly? I didn't want to handle all other structured cases with Transpose/Adjoint anyway, so at best these cases here. I'm fine with reducing the code as you propose.

stevengj replied:
> So you want to not treat the non-real transpose case explicitly?

I don't think it's worth the code bloat; we can always add an optimized method later if there is a demonstrated need.

stevengj replied:
Can be slightly generalized to:
```jl
dot(x::AbstractVector, transA::Transpose{<:Real}, y::AbstractVector) = adjoint(dot(y, transA.parent, x))
```

dkarrasch replied:
So I went with your short solutions. Many thanks!

thchr commented:
Given the proposed function signature, this should probably use `eachindex` and `axes` rather than `1:n` and `size`; otherwise, access will in fact not be inbounds if supplied with e.g. an `OffsetArray`, leading to arbitrary answers/segfaults.
For the general signature, the fix could be something like:
```julia
function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
    (axes(x)..., axes(y)...) == axes(A) || throw(DimensionMismatch())

    T = promote_type(promote_type(eltype(x), eltype(A)), eltype(y))
    s = zero(T)
    @inbounds for j in eachindex(y)
        yj = y[j]
        temp = zero(T)
        @simd for i in eachindex(x)
            temp += adjoint(x[i]) * A[i,j]
        end
        s += temp * yj
    end
    return s
end
```
(The above also removes some of the early-termination calls; they don't really seem warranted to me)

Irregardless of the above, I noticed that this is not doing better than ordinary `x'*A*y`, time-wise, for large input. Specifically, if 
```julia
x′ = rand(N); A′ = rand(N,M); y′ = rand(M)
```
Then, for `N =  M = 1000`
```julia
@btime adjoint($x′)*$A′*$y′  # 193.201 μs (1 allocation: 7.94 KiB)
@btime dot($x′,$A′,$y′)      # 312.699 μs (0 allocations: 0 bytes)
```
Though it does better for `(N,M)` less than ~700. Still, there might be room for improvement.

<details><summary>More timings ...</summary>

For `N = M = 2000`:
```julia
@btime adjoint($x′)*$A′*$y′  # 1.340 ms (1 allocation: 15.75 KiB)
@btime dot($x′,$A′,$y′)      # 1.615 ms (0 allocations: 0 bytes)
```
For `N =  M = 4000`
```julia
@btime adjoint($x′)*$A′*$y′  # 6.418 ms (2 allocations: 31.33 KiB)
@btime dot($x′,$A′,$y′)      # 7.083 ms (0 allocations: 0 bytes)
```

</details>

thchr commented on stdlib/LinearAlgebra/src/matmul.jl:
```diff
@@ -3,6 +3,7 @@
 # matmul.jl: Everything to do with dense matrix multiplication
 
 matprod(x, y) = x*y + x*y
+gendot(x, A, y) = adjoint(x)*A*y + adjoint(x)*A*y
```
This doesn't seem to end up being used anywhere; is it intended for use with `promote_op` in the same was as `matprod` is used?

dkarrasch replied:
Oh, this is a left-over from earlier approaches. I don't use it because for block matrices/vectors, you can't construct `zero` for the eltypes. I'll remove this.

dkarrasch commented:
@thchr Thanks for your comments. You're right, the generic one will need careful optimization, because it competes with BLAS multiplications, which has some overhead due to the allocation of the intermediate result. For very large problems, as you show, this does not outweigh the performance advantage of BLAS multiplication. So, we will need to make sure we catch up with the pure Julia multiplication.

dkarrasch committed:
improve generic dot, add tests

dkarrasch commented:
I got started with some tests for the generic case. More tests will follow, and maybe the structured versions also need a little update. I was including the above comments into the generic version. This is how it looks like now:
```julia
function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
    (axes(x)..., axes(y)...) == axes(A) || throw(DimensionMismatch())
    T = typeof(dot(first(x), first(A), first(y)))
    zeroxA = zero(adjoint(first(x))*first(A))
    s = zero(T)
    @inbounds for j in eachindex(y)
        yj = y[j]
        if !iszero(yj)
            temp = zeroxA
            @simd for i in eachindex(x)
                temp += adjoint(x[i]) * A[i,j]
            end
            s += temp * yj
        end
    end
    return s
end
```
There is a little issue with this. I wrote it to make it work for both number and matrix eltypes. In the block matrix case, the generic 3-arg dot allocates each time it adds something to `temp`. I thought I'll add a corresponding version where the eltypes of x, A, and y are `Number`s, but this introduces ambiguity. So I could restrict to number types and simplify the type stuff, or need to find a way to distinguish these cases and have `temp += adjoint(x[i]) * A[i,j]` not allocate in the block matrix case.

dkarrasch committed:
fix typos, optimize *diag, require_one_based_indexing

dkarrasch committed:
add tests

dkarrasch committed:
fix typos in triangular and tridiag

dkarrasch committed:
fix BigFloat tests in triangular

dkarrasch committed:
add sparse tests (and minor fix)

dkarrasch commented:
Tests are passing! I won't have much/any time the upcoming weeks (due to exams, holidays 🏖 etc), so I'd like to suggest to keep this PR at the current level (modulo minor improvements). The tests cover the number eltype cases and pass. I think I wrote the code such that it should work with block matrices/vectors. However, I'd like to keep testing (and potentially required adjustments/fixes) for an independent PR. Compared to my initial commits, I have also taken care of minimizing memory access whenever possible.

Regarding the fully dense version and the comparison to plain multiplication, I'm fine with putting `x'A*y` as a fallback for now, or a simple size-dependent distinction. The goal here was to introduce `dot` with the 3-arg signature, to direct `*(::Adjoint,::Matrix,::Vector)` to it, and to provide implementations for the sparse and structured cases, for which interest was expressed by different people. Optimizing the fully dense version may require thorough investigation and benchmarking, and seems to me like warrant a separate PR.

If there is no hurry, I'm also fine with developing the mentioned features in this PR further as time permits.

stevengj commented:
I think if the user calls `dot(x, A, y)`, then we should assume that they want an allocation-free algorithm even if it is slightly slower.

thchr commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -793,6 +793,33 @@ function dot(x::AbstractArray, y::AbstractArray)
+function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+    (axes(x)..., axes(y)...) == axes(A) || throw(DimensionMismatch())
+    T = typeof(dot(first(x), first(A), first(y)))
+    zeroxA = zero(adjoint(first(x))*first(A))
```
Something like 
```
T = promote_op(dot, eltype(a), eltype(A), eltype(b))
```
and similar for `zeroxA` might be better since that gets inferred (the current formulation isn't, at least on v1.1.1; appears to incur some array-size testing so far as I can tell from `@code_warntype`).

dkarrasch replied:
Yes, I know. These two lines are written such that the code can handle block matrices/vectors, so eltypes like `Vector` or `Matrix`. The issue is that calling `one` or `zero` on them fails, because the type does not contain any information on the size. And also, the same "`first`" approach is taken in the
```julia
function dot(x::AbstractArray, y::AbstractArray)
```
method a few lines above. Of course, this approach still breaks down when we have blocks of varying lengths, so this may not be a good idea either. To handle that, we would need to initialize `temp` in every loop separately. I'll push a suggestion for how to fix that.

dkarrasch committed:
handle block arrays of varying lengths

dkarrasch committed:
make generalized dot act recursively

dkarrasch committed:
add generalized dot for symmetric/Hermitian matrices

dkarrasch committed:
fix triangular case

dkarrasch committed:
more complete tests for Symmetric/Hermitian

dkarrasch committed:
fix UnitLowerTriangular case

dkarrasch committed:
fix complex case in symmetric gendot

dkarrasch committed:
interpret dot(x, A, y) as dot(A'x, y), test accordingly

dkarrasch committed:
use correct tolerance in triangular tests

dkarrasch commented:
The one failing test "wasn't me". So this seems to be in rather good shape, I'd say. Is it realistic to include this into Julia v1.3? I think it would be nice to have that feature (if it is complete), and optimize it later if necessary.

I have consistently interpreted `dot(x, A, y)` as `dot(A'x, y)`, which in most cases is nothing but `dot(x, A*y)` anyway, except for non-associative numbers. I felt that the interpretation as `dot(A'x, y)` really expresses what we're doing: (i) computing some `dot`, which acts recursively, in contrast to ternary `*`, (ii) we first multiply `x` and `A`, and only then with elements from `y`, (iii) but makes obvious what the implementation difference must be to `dot(A'x, y)`, i.e., not storing `A'x`.

I have covered all special matrices (bidiag, tridiags, diagonal, uniformscaling, symmetric/hermitian, triangular), what is missing is `(Upper)Hessenberg`, if I haven't overlooked anything else. Tests cover the `Number` eltype cases, the generic version (but no other) is also tested for block matrices. The hermitian/symmetric case is slightly suboptimal in that we go over the diagonal twice too often, but I felt that writing out all 4 cases with `:U`, `:L`, `x===y` and `x !== y` would lead to "code bloat", as @stevengj would say.

dkarrasch committed:
add gendot for UpperHessenberg, and tests

dkarrasch commented:
I'm close to pulling the plug for a while, so don't be surprised to see no response. `UpperHessenberg` now also has its own ternary dot method.

I'd be very happy to find the feature in Julia v1.3, of which feature freeze is in one week, or so?

StefanKarpinski commented:
Correct: feature freeze is slated for Aug 15th.

antoine-levitt commented:
> I think if the user calls dot(x, A, y), then we should assume that they want an allocation-free algorithm even if it is slightly slower.

Not sure this is a correct assumption; If I see that there's a specialized implementation for this, I'm going to assume that it's faster than the standard ones, or people wouldn't have bothered writing a specialized implementation. Especially in the common case where x and y are vectors and A is a dense matrix, the allocation of a temporary vector doesn't matter much. Also BLAS might use threading and therefore be (potentially much) faster. At least if this method has the potential to be slower than `dot(x,A*y)` it should be documented as such.

stevengj commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -793,6 +793,36 @@ function dot(x::AbstractArray, y::AbstractArray)
+without storing the intermediate result of `A'x`. As for the two-argument
+[`dot(_,_)`](@ref), this acts recursively. Moreover, for complex vectors, the
+first vector is conjugated.
+"""
```
This docstring should presumably be attached to `function dot end` and not to the `Number` method.

Also, it seems a bit weird to me to define it as `dot(A'x, y)` rather than `dot(x, A*y)`.    The fact that these two *must* be equivalent is the *definition* of [the adjoint `A'`](https://en.wikipedia.org/wiki/Hermitian_adjoint), so it seems safe to me to use the simpler definition `dot(x, A*y)`.

If you have a non-associative `*` algebra, it seems like you would need to define `A'` such that  `dot(A'x, y) == dot(x, A*y)` still holds.

dkarrasch replied:
Yes, of course. I was stretching things a bit. What I wanted to emphasize is the order of multiplication in the implementation. I’ll change the docstring accordingly.

stevengj commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -398,3 +398,5 @@ Array(s::UniformScaling, dims::Dims{2}) = Matrix(s, dims)
 Diagonal{T}(s::UniformScaling, m::Integer) where {T} = Diagonal{T}(fill(T(s.λ), m))
 Diagonal(s::UniformScaling, m::Integer) = Diagonal{eltype(s)}(s, m)
+
+dot(x::AbstractVector, J::UniformScaling, y::AbstractVector) = J.λ * dot(x, y)
```
This assumes `*` is commutative… Probably this should be a specialized method for `UniformScaling{<:Union{Real,Complex}}`.

dkarrasch replied:
Right. I’ll add a generic `UniformScaling` version similar to the one for `Diagonal`.

dkarrasch committed:
fix docstring of 3-arg dot

dkarrasch committed:
add generic 3-arg dot for UniformScaling

dkarrasch committed:
add generic fallback

This should be only relevant to cases like `dot(x, J, y)`, where `x` and `y` are vectors of quaternion vectors, and `J` is a quaternion `UniformScaling`.

dkarrasch committed:
add gendot with middle argument Number

dkarrasch commented:
I have tried to improve/fix a couple of things quickly via the web interface; thanks @stevengj for the review. I have also created a NEWS entry. Unfortunately, there is a little merge conflict which I don’t know how to resolve via the web interface. If someone could rebase, that would be much appreciated.

stevengj commented:
I fixed the merge conflict.

stevengj commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -872,6 +872,39 @@ function dot(x::AbstractArray, y::AbstractArray)
+[`dot(_,_)`](@ref), this acts recursively. Moreover, for complex vectors, the
+first vector is conjugated.
+"""
+function dot end
```
Actually, probably better to attach this to the generic fallback `dot(x, A, y) = dot(x, A*y)` definition, so that it doesn't conflict with the 2-argument `dot` docstring.

stevengj commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -400,3 +400,12 @@ Array(s::UniformScaling, dims::Dims{2}) = Matrix(s, dims)
+function dot(x::AbstractVector, J::UniformScaling, y::AbstractVector)
+    return mapreduce(t -> dot(t[1], J.λ, t[2]), +, zip(x, y))
+end
+function dot(x::AbstractVector, a::Number, y::AbstractVector)
```
Since `J.λ <: Number` always, it is probably better to define
```jl
dot(x::AbstractVector, J::UniformScaling, y::AbstractVector) = dot(x, J.λ, y)
```
and then define
```jl
dot(x::AbstractVector, a::Number, y::AbstractVector) =
    mapreduce(t -> dot(t[1], a, t[2]), +, zip(x, y))
dot(x::AbstractVector, a::Union{Real,Complex}, y::AbstractVector) = a*dot(x, y)
```
so that we have a single implementation of the scalar case.

stevengj commented on stdlib/LinearAlgebra/test/uniformscaling.jl:
```diff
@@ -330,4 +330,12 @@ end
+    λ = rand(-10:10)
+    J = UniformScaling(λ)
+    @test dot(x, J, y) == λ*dot(x, y)
+end
```
Probably should also have a test using the `Quaternion` type from testhelpers, similar to the tests in `LinearAlgebra/test/generic.jl`, to make sure we cover the non-commutative case.

dkarrasch replied:
I agree, but I’m hesitant to write tests blindly via the web interface. 🤣 I’ll do that later when I have my usual working environment around me.

dkarrasch replied:
Done.

dkarrasch committed:
attach docstring to generic fallback

dkarrasch committed:
simplify scalar/uniform scaling gendot

dkarrasch committed:
merge NEWS

dkarrasch committed:
use dot(A'x,y) for fallback

dkarrasch commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -880,7 +880,7 @@ without storing the intermediate result of `A*y`. As for the two-argument
 first vector is conjugated.
 """
-dot(x, A, y) = dot(x, A*y) # generic fallback for cases that are not covered by specialized methods
+dot(x, A, y) = dot(A'x, y) # generic fallback for cases that are not covered by specialized methods
```
Yes, sorry about that. I didn't consider the aspects you mentioned, but was lead by some tests in the large BLAS-array case. In case we rank speed higher than intermediate memory allocations, we would want to use `dot(A'x, y)` instead of `dot(x, A*y)`. So I thought this might be generally a good idea for a fallback, but probably isn't. I'll revert this change.

stevengj replied:
In the general case you can't really say *a priori*` which is faster because `A` could be anything, from a row-major array to a linear operator in ApproxFun.

`A'*x` is probably faster than `A*x` for dense column-major `A`, i.e. `Array`.   However, in that case we are not using the fallback anyway, right?

dkarrasch replied:
No, we don't. You're absolutely right. I got carried away during my efforts to speed up large dense matrix dot, which have failed miserably. :-(

dkarrasch committed:
use accessor functions in sparse code, generalize to Abstract..., tests

dkarrasch committed:
revert fallback definition

stevengj commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -872,6 +872,38 @@ function dot(x::AbstractArray, y::AbstractArray)
+"""
+dot(x, A, y) = dot(x, A*y) # generic fallback for cases that are not covered by specialized methods
+
+dot(x::Number, A::Number, y::Number) = conj(x) * A * y
```
Seems like this is redundant with the generic fallback.

stevengj commented on stdlib/LinearAlgebra/src/symmetric.jl:
```diff
@@ -457,6 +457,27 @@ end
+    require_one_based_indexing(x, y)
+    (length(x) == length(y) == size(A, 1)) || throw(DimensionMismatch())
+    Δ = A.uplo == 'U' ? UpperTriangular(A) : LowerTriangular(A)
+    D = Diagonal(A)
```
Not used?

dkarrasch replied:
No, this is a left-over from a previous version.

stevengj commented on stdlib/LinearAlgebra/src/symmetric.jl:
```diff
@@ -457,6 +457,27 @@ end
+    if x === y
+        r = 2real(dot(x, Δ, x))
+        @inbounds @simd for i in eachindex(x)
+            r -= dot(x[i], A[i,i], x[i])
```
This is not type-stable for a complex-Hermitian matrix, since `A[i,i] <: Complex` even though it is actually real.  Probably you want `real(A.data[i,i])` here (or `real(A[i,i])`, but that is slower because it performs additional checks on the indices.

stevengj replied:
...and then you have to `return complex(r)` so that the return value is type-stable

dkarrasch replied:
Hm, but in the case of block matrices, `A[i,i]` is `Hermitian(A.data[i,i|)`, where `A.data[i,i]` doesn't have to be `Hermitian` itself...? Then I need to symmetrize somehow?

dkarrasch replied:
I must be missing something about the `return complex(r)`. If any `eltype` is complex, then we initialize `r` as complex, and then add complex numbers to it, and finally return a complex number. `dot(x[i], A[i,i], x[i])` is not `Real`, and in fact the imaginary part is not exactly 0. I slightly adopted your loop-version below, and tested for inference and it passed. I'll push that and then we may continue based on the update, if necessary.

stevengj commented on stdlib/LinearAlgebra/src/symmetric.jl:
```diff
@@ -457,6 +457,27 @@ end
+function dot(x::AbstractVector, A::RealHermSymComplexHerm, y::AbstractVector)
+    require_one_based_indexing(x, y)
+    (length(x) == length(y) == size(A, 1)) || throw(DimensionMismatch())
+    Δ = A.uplo == 'U' ? UpperTriangular(A) : LowerTriangular(A)
```
This makes a copy of the data (at least in the complex case, to zero the imaginary parts on the diagonal).   It is also type-unstable.

Seems like you should really just write out the loops:
```
data = A.data
if A.uplo == 'U'
    r = zero(eltype(x)) * zero(eltype(A)) * zero(eltype(y))
    @inbounds for j = 1:length(y)
        r += dot(x[j], real(data[j,j]), y[j])
        @simd for i = 1:j-1
            Aij = data[i,j]
            r += dot(x[i], Aij, y[j]) + dot(x[j], conj(Aij), y[i])
        end
    end
else # A.uplo == 'L'
    ...
end
```
Not sure a specialized `x === y` case is worth it?

dkarrasch committed:
remove redundant Number version

dkarrasch committed:
write out loops in symmetric/hermitian case

dkarrasch committed:
test quaternions in uniformscaling gendot

dkarrasch commented:
Except for the one type-stability comment, I think I resolved all review comments. It remains to decide how to deal with the classic BLAS-float case. I tried to speed up the process by doing something like
```julia
for (Aj, yj) in zip(eachcol(A), y)
    r += dot(x, Aj) * yj
end
```
to use BLAS for the x-A-dot products, but that turned out to be even worse than the current generic version due to allocations of the `dot`-result. @stevengj proposed in the original issue to take blocks of 16-columns, and `dot`-reduce the 16-Vector with the corresponding part of `y`. Honestly, I don't feel like I'm the one who is going to find the optimal solution here. Apparently, for not too big problems the current version is faster than plain `dot(A'x, y)`, so this is not a trivial problem anyway. I'm happy with a comment in the docstring or whatever people feel is appropriate as a performance warning.

Another topic for future improvements is the block matrix case. As I said earlier, in this case the incremental update of the `temp` variable (that thing that needs to be `dot`ted with `yj`) should be in-place, so one would need a distinction of those cases. As a basis for the discussion, let me drop a version that is faster and has much less allocations than both the current version of 3-arg `dot` and `dot(x, A*y)`:
```julia
function blockdot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
    (axes(x)..., axes(y)...) == axes(A) || throw(DimensionMismatch())
    T = typeof(dot(first(x), first(A) * first(y)))
    s = zero(T)
    i₁ = first(eachindex(x))
    x₁ = first(x)
    @inbounds for j in eachindex(y)
        yj = y[j]
        if !iszero(yj)
            temp = zero(A[i₁,j]'x₁)
            @simd for i in eachindex(x)
                temp .+= adjoint(A[i,j]) * x[i]
            end
            s += dot(temp, yj)
        end
    end
    return s
end
```
This obviously fails in the classic case. I'd be very happy if we could finalize this PR within its current scope, and leave BLAS and block matrices for another PR (and another contributor 😛).

dkarrasch committed:
fix uniformscaling test

fredrikekre reviewed (approved):
Can you add a Julia 1.4 compat note to the docstring?

dkarrasch committed:
add compat note and jldoctest

dkarrasch commented:
> Can you add a Julia 1.4 compat note to the docstring?

Yes, I have done that. I have also added a little example/doctest to the docstring. While writing the doctest I was wondering if it makes sense to allow `x⋅A⋅y` as a shorthand for `dot(x, A, y)`? We have that kind of ternary-parsing for `*`, but for `⋅` this currently errors. I couldn't find where one would declare that.

I should also mention that I have explicitly **not** directed `*(::Adjoint{Vector}, ::Matrix, ::Vector})` to ternary dot, because the latter operates recursively.

For that reason, BTW, I believe
https://github.com/JuliaLang/julia/blob/8f575c5d5a7f575930eace90d261970801707a8d/stdlib/LinearAlgebra/src/adjtrans.jl#L245
is correct only for `Number` eltypes, and should be actually defined along the lines of
https://github.com/JuliaLang/julia/blob/8f575c5d5a7f575930eace90d261970801707a8d/stdlib/LinearAlgebra/src/adjtrans.jl#L247-L251
in general. Currently, you cannot multiply the adjoint of a vector `x` of 2-vectors with a vector `y` of 2x2 matrices.

StefanKarpinski commented:
Man, who knew that was going to take so long? Thanks for sticking it out, @dkarrasch!

pkofod commented:
This request originally came from @cortner's work to add preconditioners to Optim.jl! Can't believe that was in 2016.

dkarrasch commented:
Well, thanks, and thanks also to the reviewers for their patience and care, especially @stevengj!
